ykrt/src/mt.rs

```diff
@@ -347,12 +355,76 @@ impl MTThread {
                     }
                 }
             }
-            PHASE_TRACING => {
+            PHASE_TRACING | PHASE_TRACING_LOCK => {
+                if lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                    // PHASE_TRACING_LOCK requires special handling, because another thread might
+                    // have grabbed it on this Location even though this thread is tracing this
+                    // location.
+
+                    if self.inner.tracer.is_none() {
+                        // If this thread isn't tracing anything, there's little point in spinning
+                        // as the thread tracing this location is still plausibly alive.
+                        return None;
+                    }
+
+                    // We wait to grab the lock.
+                    // FIXME: this is a terrible spinlock.
+                    while lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                        lp = loc.state.load(Ordering::Acquire);
+                        yield_now();
+                    }
+
+                    if lp & PHASE_TAG != PHASE_TRACING {
+                        // While waiting for the Location to move beyond PHASE_TRACING_LOCK, we
+                        // probably raced with the thread tracing this location.
+                        return None;
+                    }
+                }
+
+                // Before we can do anything with this phase's value, we need to transition to
+                // `PHASE_TRACING_LOCK`.
+                loop {
+                    let old_lp = lp;
+                    lp = loc
+                        .state
+                        .compare_and_swap(lp, PHASE_TRACING_LOCK, Ordering::Acquire);
+                    if lp == old_lp {
+                        break;
+                    }
+                    if self.inner.tracer.is_none() {
```

Comment:
Could we have checked that earlier?

Reply:
Yes -- and we do! But we have to recheck it again after grabbing the lock.

Reply:
`self.inner.tracer` is local to this thread and I don't see anywhere in the `PHASE_TRACING` or `PHASE_TRACING_LOCK` path that would change this field. This check is also common to both cases. Seems like we should be able to check once the first thing we do?

Going by my track record of this PR though, I will probably be wrong :P

Reply:
In both instances, what's happened is that we've encountered a "we wanted to change state but failed", so the question is "should we spin or give up?" If this thread isn't tracing anything, "give up"is the better answer IMHO. Both instances are independent, so I don't think getting rid of one `self.inner.tracer` check makes much sense.

Reply:
I'm not saying get rid of it, I'm saying, unless I'm mistaken, you can check it once before all of this dance. It's local to the thread, so no other thread could change it behind our back.

Reply:
It's not that sort of check: it's only used to decide "a state change failed, should I spin or continue?" Conceptually, if I had 20 state changes in this phase, I'd make the same `self.inner.tracer` check after each.